File: experiment/ccms/src/data/block/cms_brcb.c
#include "data/block/cms_brcb.h"
#include "data/common/cms_object_name.h"

/* ---- internal stream version ---- */

int cms_brcb_encode_stream(per_stream_t *s, const cms_brcb_t *v)
{
    cms_visible_string_encode_stream(s, v->rptID, 129);
    cms_boolean_encode_stream(s, v->rptEna);
    cms_object_reference_encode_stream(s, v->datSet);
    cms_int32u_encode_stream(s, v->confRev);
    cms_rcb_opt_flds_encode_stream(s, v->optFlds);
    cms_int32u_encode_stream(s, v->bufTm);
    cms_int16u_encode_stream(s, v->sqNum);
    cms_trigger_conditions_encode_stream(s, v->trgOps);
    cms_int32u_encode_stream(s, v->intgPd);
    cms_boolean_encode_stream(s, v->gi);
    cms_boolean_encode_stream(s, v->purgeBuf);
    cms_entry_id_encode_stream(s, v->entryID);
    cms_binary_time_encode_stream(s, v->timeOfEntry_ms, v->timeOfEntry_days);
    /* resvTms OPTIONAL */
    per_encode_optional(s, v->resvTms_present);
    if (v->resvTms_present)
        cms_int16_encode_stream(s, v->resvTms);
    /* owner OPTIONAL */
    per_encode_optional(s, v->owner_present);
    if (v->owner_present)
        cms_octet_string_encode_stream(s, v->owner, v->owner_len, 64);
    return CMS_OK;
}

int cms_brcb_decode_stream(per_stream_t *s, cms_brcb_t *v)
{
    cms_visible_string_decode_stream(s, v->rptID, 129);
    cms_boolean_decode_stream(s, &v->rptEna);
    cms_object_reference_decode_stream(s, v->datSet);
    cms_int32u_decode_stream(s, &v->confRev);
    cms_rcb_opt_flds_decode_stream(s, v->optFlds);
    cms_int32u_decode_stream(s, &v->bufTm);
    cms_int16u_decode_stream(s, &v->sqNum);
    cms_trigger_conditions_decode_stream(s, v->trgOps);
    cms_int32u_decode_stream(s, &v->intgPd);
    cms_boolean_decode_stream(s, &v->gi);
    cms_boolean_decode_stream(s, &v->purgeBuf);
    cms_entry_id_decode_stream(s, v->entryID);
    cms_binary_time_decode_stream(s, &v->timeOfEntry_ms, &v->timeOfEntry_days);
    /* resvTms OPTIONAL */
    v->resvTms_present = per_decode_optional(s);
    if (v->resvTms_present)
        cms_int16_decode_stream(s, &v->resvTms);
    /* owner OPTIONAL */
    v->owner_present = per_decode_optional(s);
    if (v->owner_present)
        cms_octet_string_decode_stream(s, v->owner, &v->owner_len, 64);
    return CMS_OK;
}

/* ---- public buffer version ---- */

CMS_EXPORT int cms_brcb_encode(const cms_brcb_t *v, uint8_t *b, int *l)
{
    per_stream_t w;
    per_stream_init_write(&w, b, (size_t)*l);
    cms_brcb_encode_stream(&w, v);
    *l = (int)per_stream_bytes_written(&w);
    return CMS_OK;
}

CMS_EXPORT int cms_brcb_decode(const uint8_t *b, int l, cms_brcb_t *v)
{
    per_stream_t r;
    per_stream_init_read(&r, b, (size_t)l);
    cms_brcb_decode_stream(&r, v);
    return CMS_OK;
}

File: jcms/jcms-core/src/main/java/com/ysh/jcms/data/block/CmsBrcb.java
package com.ysh.jcms.data.block;

import com.ysh.jcms.core.CmsSequence;
import com.ysh.jcms.core.InnerField;
import com.ysh.jcms.data.*;
import com.ysh.jcms.data.common.*;
import com.ysh.jcms.data.scalar.*;
import com.ysh.jcms.data.string.CmsOctetString;
import com.ysh.jcms.data.string.CmsString;
import com.ysh.jcms.data.time.CmsBinaryTime;

public class CmsBrcb extends CmsSequence {
    @InnerField public CmsString rptID;
    @InnerField public CmsBoolean rptEna;
    @InnerField public CmsObjectReference datSet;
    @InnerField public CmsInt32U confRev;
    @InnerField public CmsRcbOptFlds optFlds;
    @InnerField public CmsInt32U bufTm;
    @InnerField public CmsInt16U sqNum;
    @InnerField public CmsTriggerConditions trgOps;
    @InnerField public CmsInt32U intgPd;
    @InnerField public CmsBoolean gi;
    @InnerField public CmsBoolean purgeBuf;
    @InnerField public CmsEntryId entryID;
    @InnerField public CmsBinaryTime timeOfEntry;
    @InnerField(optional = true) public CmsInt16 resvTms;
    @InnerField(optional = true) public CmsOctetString owner;

    public CmsBrcb() { super(new InnerBRCB()); }

    public CmsBrcb rptID(String v) { this.rptID.value(v); return this; }
    public CmsBrcb rptEna(boolean v) { this.rptEna.value(v); return this; }
    public CmsBrcb datSet(String v) { this.datSet.value(v); return this; }
    public CmsBrcb confRev(long v) { this.confRev.value(v); return this; }
    public CmsBrcb optFlds(CmsRcbOptFlds v) { this.optFlds.packed(v); return this; }
    public CmsBrcb bufTm(long v) { this.bufTm.value(v); return this; }
    public CmsBrcb sqNum(int v) { this.sqNum.value(v); return this; }
    public CmsBrcb trgOps(CmsTriggerConditions v) { this.trgOps.packed(v); return this; }
    public CmsBrcb intgPd(long v) { this.intgPd.value(v); return this; }
    public CmsBrcb gi(boolean v) { this.gi.value(v); return this; }
    public CmsBrcb purgeBuf(boolean v) { this.purgeBuf.value(v); return this; }
    public CmsBrcb entryID(byte[] v) { this.entryID.value(v); return this; }
    public CmsBrcb resvTms(int v) { this.resvTms.value(v); setPresent("resvTms", true); return this; }
    public CmsBrcb owner(byte[] v) { this.owner.value(v); setPresent("owner", true); return this; }
}
